datafusion/core/src/physical_optimizer/pruning.rs

```diff
@@ -182,6 +191,179 @@ pub trait PruningStatistics {
 /// ```
 ///
 /// See [`PruningPredicate::try_new`] and [`PruningPredicate::prune`] for more information.
+///
+/// # Background
+///
+/// ## Boolean Tri-state logic
+///
+/// To understand the details of the rest of this documentation, it is important
+/// to understand how the tri-state boolean logic in SQL works. As this is
+/// somewhat esoteric, we review it here.
+///
+/// SQL has a notion of `NULL` that represents the value is `“unknown”` and this
+/// uncertainty propagates through expressions. SQL `NULL` behaves very
+/// differently than the `NULL` in most other languages where it is a special,
+/// sentinel value (e.g. `0` in `C/C++`). While representing uncertainty with
+/// `NULL` is powerful and elegant, SQL `NULL` s are often deeply confusing when
+/// first encountered as they behave differently than most programmers may
+/// expect.
+///
+/// In most other programming languages,
+/// * `a == NULL` evaluates to `true` if `a` also had the value `NULL`
+/// * `a == NULL` evaluates to `false` if a has any other value
+///
+/// However, in SQL `a = NULL` **always** evaluates to `NULL` (never `true` or `false`):
+///
+/// | Expression     | Result   |
+/// | ------------- | --------- |
+/// | `1 = NULL`    | `NULL`    |
+/// | `NULL = NULL` | `NULL`    |
+///
+/// Also important is how `AND` and `OR` works with tri-state boolean logic as
+/// (perhaps counterintuitively) the result is **not** always NULL. While
+/// consistent with the notion of `NULL` representing “unknown”, this is again,
+/// often deeply confusing 🤯 when first encountered.
+///
+/// | Expression       | Result    | Intuition   |
+/// | ---------------  | --------- | ----------- |
+/// | `NULL AND true`  |   `NULL`  | The `NULL` stands for “unknown” and if it were `true` or `false` the overall expression value could change |
+/// | `NULL AND false` |  `false`  | If the `NULL` was either `true` or `false` the overall expression is still `false` |
+/// | `NULL AND NULL`  | `NULL`    |            |
+///
+/// | Expression      | Result    | Intuition |
+/// | --------------- | --------- | ---------- |
+/// | `NULL OR true`  | `true`    |  If the `NULL` was either `true` or `false` the overall expression is still `true` |
+/// | `NULL OR false` | `NULL`    |  The `NULL` stands for “unknown” and if it were `true` or `false` the overall expression value could change |
+/// | `NULL OR NULL`  |  `NULL`   |            |
+///
+/// ## SQL Filter Semantics
+///
+/// The SQL `WHERE` clause has a boolean expression, often called a filter or
+/// predicate. The semantics of this predicate are that the query evaluates the
+/// predicate for each row in the input tables and:
+///
+/// * Rows that evaluate to `true` are returned in the query results
+/// * Rows that evaluate to `false` are not returned (“filtered out” or “pruned” or “skipped”).
+/// * Rows that evaluate to `NULL` are **NOT** returned (also “filtered out”) – *this property appears many times in the discussion below*
```

Comment:
> Rows that evaluate to `NULL` are **NOT** returned (also “filtered out”)

Seems like this is the opposite of what discussed below. 

For example, the quote below says -- "containers that evaluate to `NULL` are returned (also 'kept')"

> /// * `NULL`: there MAY be rows that pass the predicate, **KEEPS** the container
///           Note that rewritten predicate can evaluate to NULL when some of
///           the min/max values are not known.

Did I misunderstand something here? 🤔

Reply:
This is a very good observation.

Yes the NULL semantics in filter expressions is the opposite of what the rewritten pruning predicate does with NULL (which is quite confusing). I added a note trying to clarify this: 74261e8e3